### PR TITLE
DM-44462: Try to trap case where another eups process deletes the db file

### PR DESCRIPTION
**Pull request by timj**

### CODE REVIEW

File: python/eups/stack/ProductStack.py [hunk 1]
```diff
@@ -690,9 +697,8 @@ def reload(self, flavors=None, persistDir=None, verbose=0):
         for flavor in flavors:
             fileName = self._persistPath(flavor,persistDir)
             self.modtimes[fileName] = os.stat(fileName).st_mtime
-            fd = open(fileName, "rb")
-            lookup = pickle.load(fd)
-            fd.close()
+            with open(fileName, "rb") as fd:
```
Comment (timj) [inline python/eups/stack/ProductStack.py:700]: The code is littered with the old style opens. I have only fixed the one that was in the same file as I was editing.

File: python/eups/stack/ProductStack.py [hunk 2]
```diff
@@ -252,8 +252,15 @@ def save(self, flavors=None, dir=None):
             raise CacheOutOfSync(outofsync)
 
     def _cacheFileIsInSync(self, file):
-        return (file not in self.modtimes or
-                os.stat(file).st_mtime <= self.modtimes[file])
+        if file not in self.modtimes:
+            return True
+        try:
+            older = os.stat(file).st_mtime <= self.modtimes[file]
```
Comment (ktlim) [inline python/eups/stack/ProductStack.py:258]: There seem to be a number of similar `os.stat()` calls.  They don't need protection as well?

It's curious that the only thing that seems to remove this file is an admin command.

Comment (timj) [reply python/eups/stack/ProductStack.py:258]: I will take a look at the other places.

I was also wondering how the file disappears even in the case where 30 EUPS processes in parallel are trying to read the same database file. The atomic rename should not result in the file disappearing so I realize that this is papering over the cracks a little. @RobertLuptonTheGood  do you know of a way that the pickle file could be deleted in normal usage?

File: python/eups/stack/ProductStack.py (hunk 1 above)
Comment (timj) [reply python/eups/stack/ProductStack.py:707]: I'm not sure what I'm supposed to do about the race in this code. Maybe it's a bit early in the day for me but that code gets the file name from the flavor, calls stat without checking if the cache file is there, then tries to open it. I think the code is assuming that the `reload` call is gated by a check higher up.

File: python/eups/stack/ProductStack.py (hunk 2 above)
Comment (timj) [reply python/eups/stack/ProductStack.py:258]: I've added a couple more traps but `reload` is a problem.

**Review by ktlim — APPROVED** [review]
Even if this isn't the ultimate solution, it still looks better than what was there before.

**Pull request merged**